### PR TITLE
[`type_repetition_in_bounds`]: respect MSRV for combining bounds

### LINKED ISSUE #8772: type_repetition_in_bounds should pay attention to msrv when ?Sized bound is involved
### Lint Name

type_repetition_in_bounds

### Reproducer

```toml
# .clippy.toml

msrv = "1.14.0"
```

```rust
// lib.rs

pub trait Trait<X, Y, Z> {}

pub fn f<T: ?Sized, U>(arg: usize)
where
    T: Trait<Option<usize>, Box<[String]>, bool> + 'static,
    U: Clone + Sync + 'static,
{}
```

```console
error: this type has already been used as a bound predicate
 --> src/lib.rs:7:5
  |
7 |     T: Trait<Option<usize>, Box<[String]>, bool> + 'static,
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: `-D clippy::type-repetition-in-bounds` implied by `-D clippy::pedantic`
  = help: consider combining the bounds: `T: Sized + Trait<Option<usize>, Box<[String]>, bool>`
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#type_repetition_in_bounds
```

Clippy prefers I guess one of the following signatures instead:

```rust
pub fn f<T, U>(arg: usize)
where
    T: ?Sized + Trait<Option<usize>, Box<[String]>, bool> + 'static,
    U: Clone + Sync + 'static

// or:

pub fn f<T: ?Sized + Trait<Option<usize>, Box<[String]>, bool> + 'static, U>(arg: usize)
where
    U: Clone + Sync + 'static
```

The first one is great on modern compilers, but is a parse error on Rust 1.14 and older. Meanwhile, the second is not an improvement in my opinion. The bounds next to a type parameter (as opposed to in a where-clause) are only meant to be used for small/simple bounds.

I believe clippy should not trigger type_repetition_in_bounds if the msrv is declared as 1.14.0 or older and the type parameter has a single bound of `?Sized`, with all the rest of the bounds being in the where-clause.

### Version

```text
rustc 1.62.0-nightly (7c4b47696 2022-04-30)
binary: rustc
commit-hash: 7c4b47696907d64eff5621a64eb3c6e795a9ec77
commit-date: 2022-04-30
host: x86_64-unknown-linux-gnu
release: 1.62.0-nightly
LLVM version: 14.0.1
```

### Additional Labels

_No response_

### PR DESCRIPTION
**Pull request by y21**
Fixes #8772.

changelog: [`type_repetition_in_bounds`]: respect msrv for combining `?Sized` bound

### CODE REVIEW

File: clippy_utils/src/msrvs.rs [hunk 1]
```diff
@@ -47,6 +47,7 @@ msrv_aliases! {
     1,18,0 { HASH_MAP_RETAIN, HASH_SET_RETAIN }
     1,17,0 { FIELD_INIT_SHORTHAND, STATIC_IN_CONST, EXPECT_ERR }
     1,16,0 { STR_REPEAT }
+    1,15,0 { COMBINED_MAYBE_BOUND }
```
Comment (giraffate) [inline clippy_utils/src/msrvs.rs:50]: From 1.15.0, it supports `?Sized` in where clause, so can we change this name to `sized_in_where` or `maybe_bound_in_where` or etc.?
https://github.com/rust-lang/rust/pull/37791

**Review by giraffate — COMMENTED** [review]
Can we update `clippy_lints/src/utils/conf.rs`?
https://github.com/rust-lang/rust-clippy/blob/master/book/src/development/adding_lints.md#specifying-the-lints-minimum-supported-rust-version-msrv

File: clippy_utils/src/msrvs.rs (hunk 1 above)
Comment (y21) [reply clippy_utils/src/msrvs.rs:50]: yep let's do that

**Comment by giraffate** [comment]
@bors r+

Thanks!

**Pull request merged**